### PR TITLE
Refactor read.c with avifTileGroup

### PR DESCRIPTION
**Pull request by y-guyon**
Add avifImageCopySamples().
Add AVIF_CHECKRES().
Add AVIF_ENCODER_NO_CHANGE.
Refactor write.c.

This should be easier to add more items besides primary color and auxiliary alpha in the future, such as in https://github.com/AOMediaCodec/libavif/pull/1215.

### CODE REVIEW

File: src/write.c [hunk 1]
```diff
@@ -481,7 +481,6 @@ static void avifEncoderBackupSettings(avifEncoder * encoder)
 static avifBool avifEncoderDetectChanges(const avifEncoder * encoder, avifEncoderChanges * encoderChanges)
 {
     const avifEncoder * lastEncoder = &encoder->data->lastEncoder;
-    *encoderChanges = 0;
```
Comment (jzern) [inline src/write.c:484]: It may be better to leave this and document that the value will always be set meaningfully. This is a minor change could be its own PR to keep the review size down in this one.

Comment (wantehchang) [reply src/write.c:484]: `avifEncoderDetectChanges()` follows a common convention that output parameters are set when the function succeeds. In this convention it is unspecified whether output parameters are set when the function fails, and the caller should only use the outputs after a successful call.

So the caller of `avifEncoderDetectChanges()` does not need to initialize the output before calling `avifEncoderDetectChanges()`.

File: src/read.c [hunk 2]
```diff
@@ -1375,72 +1378,34 @@ static avifBool avifDecoderDataFillImageGrid(avifDecoderData * data,
...
-                        &dstImage->yuvPlanes[AVIF_CHAN_V][(uvColOffset * pixelBytes) + ((uvRowOffset + j) * dstImage->yuvRowBytes[AVIF_CHAN_V])];
-                    memcpy(dstV, srcV, uvRowBytes);
-                }
+            const avifCropRect srcViewRect = { 0, 0, dstViewRect.width, dstViewRect.height };
+            avifImage * srcView = avifImageCreateEmpty();
+            avifImage * dstView = avifImageCreateEmpty();
+            if ((avifImageSetViewRect(dstView, dstImage, &dstViewRect) != AVIF_RESULT_OK) ||
+                (avifImageSetViewRect(srcView, tile->image, &srcViewRect) != AVIF_RESULT_OK) ||
+                (avifImageCopySamples(dstView, srcView, alpha ? AVIF_PLANES_A : AVIF_PLANES_YUV) != AVIF_RESULT_OK)) {
+                assert(AVIF_FALSE);
```
Comment (jzern) [inline src/read.c:1408]: is failure impossible in this function or should this return AVIF_FALSE?

**Comment by wantehchang** [comment]
Yannis: I will review this pull request tomorrow. Sorry about the delay.

File: src/read.c [hunk 3]
```diff
@@ -3389,6 +3354,63 @@ static avifResult avifDecoderFlush(avifDecoder * decoder)
     return AVIF_RESULT_OK;
 }
 
+// Searches for the primary color item or the auxiliary alpha item.
+// Returns the target item if found, or NULL.
+static avifDecoderItem * avifDecoderFindItem(avifDecoder * decoder, avifBool alpha, uint32_t parentItemId)
```
Comment (jzern) [inline src/read.c:3359]: Strictly speaking this is separate from the introduction of avifTileGroup. Adding this on its own would have made this change more digestible.

File: src/read.c [hunk 4]
```diff
@@ -4034,137 +3934,99 @@ avifResult avifDecoderNextImage(avifDecoder * decoder)
...
                 return AVIF_RESULT_DECODE_ALPHA_FAILED;
             }
+            avifImageFreePlanes(decoder->image, AVIF_PLANES_ALL);
 
-            avifImageStealPlanes(decoder->image, srcAlpha, AVIF_PLANES_A);
+            decoder->image->width = src->width;
+            decoder->image->height = src->height;
+            decoder->image->depth = src->depth;
+        }
+#if 1
```
Comment (jzern) [inline src/read.c:4003]: leftover debug code?

File: include/avif/internal.h [hunk 5]
```diff
@@ -29,6 +29,14 @@ extern "C" {
             return ERR;       \
     } while (0)
 
+// Forward any error to the caller now or continue execution.
+#define AVIF_CHECKRES(A)                          \
+    do {                                          \
+        const avifResult avifCheckedResult = (A); \
```
Comment (wantehchang) [inline include/avif/internal.h:35]: Nit: avifCheckedResult => checkedResult

Because of libavif's naming convention, `avifCheckedResult` unfortunately looks like a global function or a type.

We can also use a shorter name such as `result__`. The trailing underscores help avoid the -Wshadow warnings.

File: src/write.c [hunk 6]
```diff
@@ -477,7 +477,7 @@ static void avifEncoderBackupSettings(avifEncoder * encoder)
 
 // This function detects changes made on avifEncoder. It returns true on success (i.e., if every
 // change is valid), or false on failure (i.e., if any setting that can't change was changed). It
-// reports detected changes in encoderChanges.
+// reports detected changes in encoderChanges, or sets it to 0 if none.
```
Comment (wantehchang) [inline src/write.c:480]: Nit/optional: Here it is clear that 0 means no detected changes. It would be more important to document that `encoderChanges` is a bitwise-OR of detected changes.

File: src/write.c [hunk 7]
```diff
@@ -886,6 +886,55 @@ static int avifQualityToQuantizer(int quality, int minQuantizer, int maxQuantize
     return quantizer;
 }
 
+// Adds the items for a single cell or a grid of cells.
+// Outputs the mainItemId, being the only cell if there is exactly one, or the grid item for multiple cells.
```
Comment (wantehchang) [inline src/write.c:890]: Nit: Change "mainItem" to "primaryItem" and "Id" to "ID" to be consistent with current names. Change all occurrences.

File: src/write.c [hunk 8]
```diff
@@ -886,6 +886,55 @@ static int avifQualityToQuantizer(int quality, int minQuantizer, int maxQuantize
...
+static avifResult avifEncoderAddImageItems(avifEncoder * encoder,
+                                           uint32_t gridCols,
+                                           uint32_t gridRows,
+                                           uint32_t gridWidth,
+                                           uint32_t gridHeight,
+                                           avifBool alpha,
+                                           uint16_t * mainItemId)
+{
+    const uint32_t cellCount = gridCols * gridRows;
+    const char * infeName = alpha ? "Alpha" : "Color";
```
Comment (wantehchang) [inline src/write.c:900]: Nit: I suggest adding a local variable:

```
    size_t infeNameSize = 6;
```

and use it instead of the expression `strlen(infeName) + 1` below.

If you strongly prefer, you can call the `strlen()` function:

```
    size_t infeNameSize = strlen(infeName) + 1;
```

But I think 6 is clearly correct if it is right after the `const char * infeName = alpha ? "Alpha" : "Color";` line.

File: src/write.c [hunk 9]
```diff
@@ -886,6 +886,55 @@ static int avifQualityToQuantizer(int quality, int minQuantizer, int maxQuantize
...
+
+        if (cellCount > 1) {
+            item->dimgFromID = *mainItemId;
+            item->hiddenImage = AVIF_TRUE;
+        } else {
+            *mainItemId = item->id;
+        }
+    }
+
+    // Returns uint16_t * mainItemId instead of avifEncoderItem * mainItem because
```
Comment (wantehchang) [inline src/write.c:933]: This comment is confusing because there is no variable named `mainItem` in this function.

**Review by wantehchang — COMMENTED** [review]
Yannis: I am still reviewing this pull request. This pull request contains some independent changes. It would be good to break it into smaller pull requests for easier code review. I don't want to make you do extra work, so I will only describe the independent pieces that I have identified.

I suggest we start with these two:
1. The `AVIF_CHECKRES` macro in internal.h.
2. The changes to write.c.

You can put these two in one pull request. I first publish my comments on these two.

File: src/write.c [hunk 10]
```diff
@@ -1092,52 +1112,17 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
...
-                        primaryItem->irefType = "prem";
-                        primaryItem->irefToID = item->id;
-                    }
-                }
+            uint16_t alphaItemId;
+            AVIF_CHECKRES(avifEncoderAddImageItems(encoder, gridCols, gridRows, gridWidth, gridHeight, /*alpha=*/AVIF_TRUE, &alphaItemId));
+            avifEncoderItem * alphaItem = avifEncoderDataFindItemByID(encoder->data, alphaItemId);
+            assert(alphaItem);
+            alphaItem->irefToID = colorItemId;
+            alphaItem->irefType = "auxl";
```
Comment (wantehchang) [inline src/write.c:1120]: Nit: I suggest setting `irefToID` and `irefType` in the same order here and at lines 1124-1125. (I don't know which order is "better".)

File: src/read.c (hunk 2 above)
Comment (y-guyon) [reply src/read.c:1408]: Failure _should_ be impossible.

File: src/read.c (hunk 4 above)
Comment (y-guyon) [reply src/read.c]: I changed the `0` to `1` to verify that the code inside worked properly but forgot to change the `1` back to `0`. Done.

File: src/write.c (hunk 7 above)
Comment (y-guyon) [reply src/write.c]: > Nit: Change "mainItem" to "primaryItem" 

If `alpha` is true, this is not the primary item that is returned, but the "top-level" one: `the only cell if there is exactly one, or the grid item for multiple cells`. That is why I named it "main" instead of "primary".

> "Id" to "ID" to be consistent with current names. Change all occurrences.

This is against Google style but it is not even applied in this project. Done.

File: src/read.c (hunk 3 above)
Comment (y-guyon) [reply src/read.c:3359]: Done in https://github.com/AOMediaCodec/libavif/pull/1298.

File: include/avif/internal.h [hunk 11]
```diff
@@ -199,6 +207,10 @@ avifBool avifImageScale(avifImage * image,
                         uint32_t imageDimensionLimit,
                         avifDiagnostics * diag);
 
+// Copies the samples from srcImage to dstImage. dstImage must be allocated.
+// srcImage and dstImage must have the same width, height, depth and subsampling.
```
Comment (wantehchang) [inline include/avif/internal.h:211]: Nit: subsampling => yuvFormat

File: src/read.c [hunk 12]
```diff
@@ -1375,72 +1378,34 @@ static avifBool avifDecoderDataFillImageGrid(avifDecoderData * data,
...
-                    memcpy(dstU, srcU, uvRowBytes);
-
-                    const uint8_t * srcV = &tile->image->yuvPlanes[AVIF_CHAN_V][j * tile->image->yuvRowBytes[AVIF_CHAN_V]];
-                    uint8_t * dstV =
-                        &dstImage->yuvPlanes[AVIF_CHAN_V][(uvColOffset * pixelBytes) + ((uvRowOffset + j) * dstImage->yuvRowBytes[AVIF_CHAN_V])];
-                    memcpy(dstV, srcV, uvRowBytes);
-                }
+            const avifCropRect srcViewRect = { 0, 0, dstViewRect.width, dstViewRect.height };
+            avifImage * srcView = avifImageCreateEmpty();
+            avifImage * dstView = avifImageCreateEmpty();
```
Comment (wantehchang) [inline src/read.c:1404]: `srcView` and `dstView` can be created outside the for loops and reused for all loop iterations.

Need to handle `avifImageCreateEmpty()` failures.

Need to destroy `srcView` and `dstView` when done.

File: src/scale.c [hunk 13]
```diff
@@ -180,3 +182,33 @@ avifBool avifImageScale(avifImage * image,
 }
 
 #endif
+
+avifResult avifImageCopySamples(avifImage * dstImage, const avifImage * srcImage, avifPlanesFlags planes)
```
Comment (wantehchang) [inline src/scale.c:186]: Define `avifImageCopySamples()` in src/avif.c. This function is similar to `avifImageCopy()`, so they should be defined in the same file. This function is unrelated to scaling an image.

File: src/scale.c [hunk 14]
```diff
@@ -180,3 +182,33 @@ avifBool avifImageScale(avifImage * image,
...
+{
+    for (int c = AVIF_CHAN_Y; c <= AVIF_CHAN_A; ++c) {
+        if ((!(planes & AVIF_PLANES_YUV) && (c != AVIF_CHAN_A)) || (!(planes & AVIF_PLANES_A) && (c == AVIF_CHAN_A))) {
+            continue;
+        }
```
Comment (wantehchang) [inline src/scale.c:191]: Nit/optional: This if statement (lines 189-191) is a little hard to understand. I seem to remember you used an alternative solution like the following:
```
    int firstPlane = AVIF_CHAN_A;
    int lastPlane = AVIF_CHAN_V;
    if (planes & AVIF_PLANES_YUV) {
        firstPlane = AVIF_CHAN_Y;
    }
    if (planes & AVIF_PLANES_A) {
        lastPlane = AVIF_CHAN_A;
    }
    for (int c = firstPlane; c <= lastPlane; ++c) {
```
But I can't find it in the source tree. Maybe it was deleted or rewritten during the code review of the pull request.

The alternative solution is also a little hard to understand, but it is easier to parse.

File: src/scale.c [hunk 15]
```diff
@@ -180,3 +182,33 @@ avifBool avifImageScale(avifImage * image,
...
+        const uint32_t planeWidth = avifImagePlaneWidth(srcImage, c);
+        const uint32_t planeHeight = avifImagePlaneHeight(srcImage, c);
+        const uint8_t * srcRow = avifImagePlane(srcImage, c);
+        uint8_t * dstRow = avifImagePlane(dstImage, c);
+        const uint32_t srcRowBytes = avifImagePlaneRowBytes(srcImage, c);
+        const uint32_t dstRowBytes = avifImagePlaneRowBytes(dstImage, c);
+        if ((planeWidth != avifImagePlaneWidth(dstImage, c)) || (planeHeight != avifImagePlaneHeight(dstImage, c)) ||
+            (!srcRow != !dstRow) || (srcImage->depth != dstImage->depth)) {
```
Comment (wantehchang) [inline src/scale.c:199]: Nit: Test `srcImage->depth != dstImage->depth` outside the for loop. This check is loop invariant.

File: src/scale.c [hunk 16]
```diff
@@ -180,3 +182,33 @@ avifBool avifImageScale(avifImage * image,
...
+            return AVIF_RESULT_INVALID_ARGUMENT;
+        }
+        if (!srcRow) {
+            continue;
+        }
+
+        const size_t planeWidthBytes = planeWidth * (size_t)(avifImageUsesU16(srcImage) ? 2 : 1);
```
Comment (wantehchang) [inline src/scale.c:206]: Nit: Similarly the bytes-per-pixel expression is loop invariant and can be saved in a local variable outside the for loop.

Also in `avifImageCopy()` we use this form (left shift):

```
       uint32_t planeWidthBytes = avifImagePlaneWidth(dstImage, plane) << (dstImage->depth > 8);
```

Would be good to standardize on one form (either multiplication or left shift).

**Review by wantehchang — COMMENTED** [review]
Yannis: The next independent part of this pull request is `avifImageCopySamples()` and the code that uses it. It would be good to split this part to a separate pull request.

Here are my comments on this part.

**Comment by y-guyon** [comment]
> Yannis: The next independent part of this pull request is `avifImageCopySamples()` and the code that uses it. It would be good to split this part to a separate pull request.
> 
> Here are my comments on this part.

Done in https://github.com/AOMediaCodec/libavif/pull/1300.

File: src/read.c [hunk 17]
```diff
@@ -3390,6 +3394,29 @@ static avifDecoderItem * avifDecoderDataFindItem(avifDecoderData * data, avifBoo
     return NULL;
 }
 
+static avifResult avifDecoderGenerateImageTiles(avifDecoder * decoder, avifTileGroup * tileGroup, avifDecoderItem * item, avifBool alpha)
```
Comment (jzern) [inline src/read.c:3397]: The parameter order is consistent with the file, I just noticed there's no this/in/out ordering that's common.

File: src/read.c [hunk 18]
```diff
@@ -3958,11 +3923,56 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder,
...
+            return AVIF_RESULT_DECODE_ALPHA_FAILED;
+        }
+        avifImageFreePlanes(decoder->image, AVIF_PLANES_ALL);
+
+        decoder->image->width = src->width;
+        decoder->image->height = src->height;
+        decoder->image->depth = src->depth;
+    }
+#if 0
+    // This code is currently unnecessary as the CICP is always set by the end of avifDecoderParse().
```
Comment (jzern) [inline src/read.c:3964]: is that going to change or instead of the code should just a note be left here?

**Review by jzern — APPROVED** [review]

File: src/read.c [hunk 19]
```diff
@@ -763,17 +763,22 @@ static avifDecoderItem * avifMetaFindItem(avifMeta * meta, uint32_t itemID)
     return NULL;
 }
 
+// A group of AVIF tiles, such as a single tile or a grid of multiple tiles.
+typedef struct avifTileGroup
```
[the remark this answers is not in the document]
Comment (wantehchang) [reply src/read.c:767]: Correction: The last sentence should read:

> Perhaps we can call this struct `avifItemTiles`, which means the tiles of an (image) item.

File: src/read.c (hunk 18 above)
Comment (y-guyon) [reply src/read.c:3964]: This code was only moved in this PR, not changed.

File: src/read.c [hunk 20]
```diff
@@ -1284,21 +1289,20 @@ static avifBool avifDecoderGenerateImageGridTiles(avifDecoder * decoder, avifIma
 // dstImage. Only the freshly decoded tiles are considered, skipping the already
 // copied or not-yet-decoded tiles.
 static avifBool avifDecoderDataFillImageGrid(avifDecoderData * data,
-                                             avifImageGrid * grid,
                                              avifImage * dstImage,
-                                             unsigned int firstTileIndex,
+                                             const avifTileGroup * tileGroup,
```
[the remark this answers is not in the document]
Comment (y-guyon) [reply src/read.c]: I do not follow. The name of the function describes its effect, not the arguments. I do not see a cost here. A shorter signature is even a benefit to me.

File: src/read.c [hunk 21]
```diff
@@ -763,17 +763,22 @@ static avifDecoderItem * avifMetaFindItem(avifMeta * meta, uint32_t itemID)
     return NULL;
 }
 
+// A group of AVIF tiles, such as a single tile or a grid of multiple tiles.
```
Comment (wantehchang) [inline src/read.c:766]: Add "in an image item" to the comment.

File: src/read.c [hunk 22]
```diff
@@ -1284,21 +1289,20 @@ static avifBool avifDecoderGenerateImageGridTiles(avifDecoder * decoder, avifIma
 // dstImage. Only the freshly decoded tiles are considered, skipping the already
 // copied or not-yet-decoded tiles.
 static avifBool avifDecoderDataFillImageGrid(avifDecoderData * data,
-                                             avifImageGrid * grid,
                                              avifImage * dstImage,
-                                             unsigned int firstTileIndex,
+                                             const avifItemTiles * group,
```
Comment (wantehchang) [inline src/read.c:1293]: Nit: `group` => `itemTiles` or `imageTiles` or `tiles`

Also fix other occurrences.

File: src/read.c [hunk 23]
```diff
@@ -3390,6 +3394,29 @@ static avifDecoderItem * avifDecoderDataFindItem(avifDecoderData * data, avifBoo
     return NULL;
 }
 
+static avifResult avifDecoderGenerateImageTiles(avifDecoder * decoder, avifItemTiles * group, avifDecoderItem * item, avifBool alpha)
```
Comment (wantehchang) [inline src/read.c:3397]: Optional: Hmm... Given this function name, perhaps the first struct name I suggested, `avifImageTiles`, isn't bad after all.

The reason I thought `avifImageTiles` was a bad name is that an `avifImage` may consist of two HEIF image items (color and alpha). The name `avifImageTiles` could refer to an `avifImage` or a HEIF image item. So it could be ambiguous.

File: src/read.c [hunk 24]
```diff
@@ -3617,36 +3641,14 @@ avifResult avifDecoderReset(avifDecoder * decoder)
...
-
+        AVIF_CHECKRES(avifDecoderGenerateImageTiles(decoder, &data->color, colorItem, /*alpha=*/AVIF_FALSE));
+        if ((data->color.grid.rows == 0) || (data->color.grid.columns == 0)) {
             if (colorItem->progressive) {
                 decoder->progressiveState = AVIF_PROGRESSIVE_STATE_AVAILABLE;
+                const avifTile * colorTile = &data->tiles.tile[0];
                 if (colorTile->input->samples.count > 1) {
                     decoder->progressiveState = AVIF_PROGRESSIVE_STATE_ACTIVE;
-                    decoder->imageCount = colorTile->input->samples.count;
+                    decoder->imageCount = (int)colorTile->input->samples.count;
```
Comment (wantehchang) [inline src/read.c:3651]: Nit: I usually only add this kind of cast if the compiler warns. If you prefer this cast, please also add a cast to line 3571.

File: src/read.c [hunk 25]
```diff
@@ -3958,8 +3923,53 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder,
             }
         }
 
-        ++*decodedTileCount;
+        ++group->decodedTileCount;
+    }
+    return AVIF_RESULT_OK;
+}
+
+avifResult avifDecoderOutputDecodedTiles(avifDecoder * decoder, avifItemTiles * group, avifBool alpha, unsigned int oldDecodedTileCount)
```
Comment (wantehchang) [inline src/read.c:3931]: Add `const` to `avifItemTiles * group`

File: src/read.c [hunk 26]
```diff
@@ -3958,8 +3923,53 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder,
...
+        assert(group->tileCount == (group->grid.rows * group->grid.columns));
+        AVIF_CHECKERR(avifDecoderDataFillImageGrid(decoder->data, decoder->image, group, oldDecodedTileCount, alpha),
+                      AVIF_RESULT_INVALID_IMAGE_GRID);
+        return AVIF_RESULT_OK;
+    }
+
+    // Normal (most common) non-grid path.
+    assert(group->tileCount == 1);
+    avifImage * src = decoder->data->tiles.tile[group->firstTileIndex].image;
+    const avifPlanesFlag planes = alpha ? AVIF_PLANES_A : AVIF_PLANES_YUV;
```
Comment (wantehchang) [inline src/read.c:3951]: Move this to line 3972. This variable is only used there.

File: src/read.c [hunk 27]
```diff
@@ -3958,8 +3923,53 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder,
...
+    if ((decoder->image->width != src->width) || (decoder->image->height != src->height) || (decoder->image->depth != src->depth)) {
+        if (alpha) {
+            avifDiagnosticsPrintf(&decoder->diag, "decoder->image does not match srcAlpha in width, height, or bit depth");
```
Comment (wantehchang) [inline src/read.c:3954]: Nit: In the diagnostic message, `srcAlpha` needs to be updated for the new code. `srcAlpha` is a variable name in the original code.

File: src/read.c [hunk 28]
```diff
@@ -3976,128 +3986,48 @@ avifResult avifDecoderNextImage(avifDecoder * decoder)
...
-        (decoder->data->decodedAlphaTileCount != decoder->data->alphaTileCount)) {
+    if ((decoder->data->color.decodedTileCount != decoder->data->color.tileCount) ||
+        (decoder->data->alpha.decodedTileCount != decoder->data->alpha.tileCount)) {
         assert(decoder->allowIncremental);
         // The image is not completely decoded. There should be no error unrelated to missing bytes,
         // and at least some missing bytes.
-        assert((prepareColorTileResult == AVIF_RESULT_OK) || (prepareColorTileResult == AVIF_RESULT_WAITING_ON_IO));
-        assert((prepareAlphaTileResult == AVIF_RESULT_OK) || (prepareAlphaTileResult == AVIF_RESULT_WAITING_ON_IO));
-        assert((prepareColorTileResult != AVIF_RESULT_OK) || (prepareAlphaTileResult != AVIF_RESULT_OK));
+        assert((prepareColorTileResult == AVIF_RESULT_WAITING_ON_IO) || (prepareAlphaTileResult == AVIF_RESULT_WAITING_ON_IO));
```
Comment (wantehchang) [inline src/read.c:4030]: The new assertion seems different from the old assertions. Please confirm the new assertion is correct. (I didn't try to understand this change.)

**Review by wantehchang — APPROVED** [review]
LGTM. I checked the code carefully this time.

File: src/read.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (wantehchang) [reply src/read.c]: This isn't that important. I was just stating my first impression when I had only skimmed the diffs quickly. What I saw was:

Before the change:
```
static avifBool avifDecoderDataFillImageGrid(avifDecoderData * data,
                                             avifImageGrid * grid,
```

Notice how the function name "avifDecoderDataFillImageGrid" matches the types of the first two parameters. This matching is gone after the change.

Now, looking at this more, I think the original matching was actually misleading, because the "FillImageGrid" part of the function name actually refers to "fill `avifImage * dstImage`" rather than "fill `avifImageGrid * grid`".

File: src/read.c (hunk 22 above)
Comment (y-guyon) [reply src/read.c]: Done. At first I avoided that because it makes the order `tiles->grid` look weird. I also considered `item->grid` but it conflicts with `avifDecoderItem`.

File: src/read.c (hunk 24 above)
Comment (y-guyon) [reply src/read.c:3651]: I had the warning in my IDE (CLion).

File: src/read.c (hunk 28 above)
Comment (y-guyon) [reply src/read.c:4029]: It is.

File: src/read.c [hunk 29]
```diff
@@ -3958,8 +3923,53 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder,
...
+
+    // Grid path.
+    if ((tiles->grid.rows > 0) && (tiles->grid.columns > 0)) {
+        assert(tiles->tileCount == (tiles->grid.rows * tiles->grid.columns));
+        AVIF_CHECKERR(avifDecoderDataFillImageGrid(decoder->data, decoder->image, tiles, oldDecodedTileCount, alpha),
+                      AVIF_RESULT_INVALID_IMAGE_GRID);
+        return AVIF_RESULT_OK;
+    }
+
+    // Normal (most common) non-grid path.
```
Comment (wantehchang) [inline src/read.c:3948]: Nit/optional: Should we preserve the full comment from the original code?

```
// Normal (most common) non-grid path. Just steal the planes from the only "tile".
```

Either is fine by me. Just wanted to confirm you meant to delete the second sentence in the original comment.

**Review by wantehchang — APPROVED** [review]
LGTM!

File: src/read.c (hunk 24 above)
Comment (wantehchang) [reply src/read.c:3651]: Do you know why it didn't warn about line 3571 (which is exactly the same)? Does the IDE only check the code that was modified?

File: src/read.c (hunk 22 above)
Comment (wantehchang) [reply src/read.c]: I am also not very happy with `tiles`. It's hard to come up with a good, short name for this parameter.

Comment (wantehchang) [reply src/read.c]: One possibility is to name the struct `avifTileInfo` or `avifItemTileInfo`, and name the function parameter `info`.

File: src/read.c [hunk 30]
```diff
@@ -4242,23 +4172,19 @@ uint32_t avifDecoderNearestKeyframe(const avifDecoder * decoder, uint32_t frameI
...
+    if (info->decodedTileCount == 0) {
         return 0;
     }
 
-    if ((grid->rows > 0) && (grid->columns > 0)) {
-        // Grid of AVIF tiles (not to be confused with AV1 tiles).
-        const uint32_t tileHeight = decoder->data->tiles.tile[firstTileIndex].height;
-        return AVIF_MIN((decodedTileCount / grid->columns) * tileHeight, decoder->image->height);
+    if ((info->grid.rows > 0) && (info->grid.columns > 0)) {
+        // Grid of AVIF info (not to be confused with AV1 tiles).
```
Comment (wantehchang) [inline src/read.c:4185]: AVIF info => AVIF tiles

**Review by wantehchang — APPROVED** [review]
LGTM. There is a find-and-replace error. Please correct that and merge this pull request. Thank you.

**Comment by wantehchang** [comment]
Yannis: Are you planning to merge this pull request?

**Review by wantehchang — APPROVED** [review]
LGTM.
